### ISSUE TITLE
Multiple tracks and <trk> segments

### ISSUE THREAD

**Issue by velocat** [Raruto/leaflet-elevation#56]
Sometimes users can download gpx of this kind:

```html
<gpx>
      <trk> .... </trk>
      <trk> .... </trk>
</gpx>
```

It contains several tracks, and they are not necessarily an extension of each other.
The ability to create such multitracks is provided by **MapSource** and **BaseCamp** - the basic programs for working with tracks.

![Multitrack](https://user-images.githubusercontent.com/534853/83339121-48b76880-a2d3-11ea-96aa-9feeb0b83733.PNG)

It would be nice to use the **`area.defined`** feature to correctly display these codes:
https://observablehq.com/@d3/area-with-missing-data

```js
let area = this._area = d3.area().curve(interpolation)
    .defined(d => !isNaN(d.z))
    .x(d => (d.xDiagCoord = x(d.dist)))
    .y0(this._height())
    .y1(d => y(d.z));
```

But I just can’t figure out how to apply the filter to SVG, like this:

```js
svg.append("path")
 	.datum(d.z.filter(area.defined())) // d.z ??
 	.attr("fill", "#eee")
 	.attr("d", area);
```

 I just did not quite understand the principle of building SVG.
Can you tell me a thought?

In general, it seems to me that it would not be bad to add this to the general code :)

**Comment by Raruto** [comment]
Hi velocat,

> I just can’t figure out how to apply the filter to SVG, like this:

Based on your example, it could be such a thing:

```js
var filteringFunc = this._area.defined();
var filteredData = this._data.filter(filteringFunc),

svg.append("path")
  .datum(filteredData)
  .attr("fill", "#eee")
  .attr("d", area);
```

> I just did not quite understand the principle of building SVG.

Try reading this article to better understand how it works:

 https://bocoup.com/blog/showing-missing-data-in-line-charts

> In general, it seems to me that it would not be bad to add this to the general code :)

An "intermediate" solution already exists:

https://github.com/Raruto/leaflet-elevation/blob/f0a2de6353af9ad0d3786e5ba58c195e78efa95e/src/leaflet-elevation.js#L595

But at the moment this library doesn't support multi-track segments, anyway pull requests are welcome.

_Happy further investigations,
Raruto_

**Comment by velocat** [comment]
> But at the moment this library doesn't support multi-track segments, anyway pull requests are welcome.

As you can see from the picture above, I still adapted the multi-track)) like this:
```
			loadTrace = function(id) {				
				var Maintrace = {};
				
				$.ajax({
					//dataType: "xml", 
					url: "aljax2.php",
					data: {mode: "getcontent", id:id},
					success: function(xml) {
						Maintrace.gpx = new L.GPX(xml,{
							async: true,
							joinTrackSegments: true,
							joinTracks: false,
							marker_options : {
								wptIcons: arrayIcons,
								startIconUrl : null,//'images/map/icon-start.png' , 
								endIconUrl : null,//'images/map/icon-end.png' , 
								iconSize: [13, 13],
								iconAnchor: [6, 13],
								shadowUrl : '' 
							},							
							polyline_options: {
								color: colors[countColor],
								stroke: '#000',
								opacity: 0.75,
								weight: 4,
								lineCap: 'round',
							}
						}).on('loaded', function(e) {
							if(Tracks.length>1){
								controlLayer.addTo(mymap);
								controlLayer.addBaseLayer(e.target, "Все участки");	
								MainTrack_id = e.target._leaflet_id;
							}	
							var i = 0;
							for (var track in Tracks) {
								loadTrack(track, i++)
							}		
							allLayers = e.layers;
							waypoints.addTo(mymap);
						}).on("addline", function(e) {
							var line = e.line;
							line.options.distanceMarkers={lazy:true};
							if(!Maintrace.gpx.lines){
								Maintrace.gpx.lines = []
							}
							Maintrace.gpx.line = line;
							Maintrace.gpx.lines.push(line)
							Tracks.push(e.element);
						}).on('addpoint', function(e) {
							waypoints.addLayer(e.point);
						});
						traces.push(Maintrace);
					} 
				});	
			}


			function loadTrack(track, i) {
				var trace = {};
				
				var XMLTrack = xmlToString(Tracks[track]);
				
				trace.gpx = new L.GPX(XMLTrack, {
					async: true,
					index: i,
					marker_options: {
						wptIcons: arrayIcons,
						startIconUrl : 'images/map/icon-start.png' , 
						endIconUrl : 'images/map/icon-end.png' , 
						iconSize: [13, 13],
						iconAnchor: [6, 13],
						shadowUrl : '' 
					},
					polyline_options: {
						color: colors[countColor],
						stroke: '#000',
						opacity: 0.75,
						weight: 4,
						lineCap: 'round',
					}
				});

				trace.gpx.on('loaded', function(e) {
					controlLayer.addBaseLayer(e.target, e.target.get_name());
					if (e.target.options.index == 0) {
						setElevationTrack(0);
					} else {
						mymap.removeLayer(e.target);
					}
				})

				trace.gpx.on("addline", function(e) {
					var line = e.line;
					line.options.distanceMarkers={lazy:true};
					trace.gpx.line = line;
				})
				
				trace.gpx.on("click", function(e) {
					//console.log(i)
				})
				
				trace.gpx.addTo(mymap);
				traces.push(trace);
			}
```
The function that you indicated works great and that is why there is a flat line of heights from the end of one segment to the beginning of a new one. It is only necessary to paint over it :)) In general, it remains to deal with the paths in svg. If I find a solution, I will definitely share it.

**Comment by Raruto** [comment]
> As you can see from the picture above, I still adapted the multi-track like this:

I can't see it in your code, but I think you solved this problem by calling the [`addData()`](https://github.com/Raruto/leaflet-elevation/blob/f0a2de6353af9ad0d3786e5ba58c195e78efa95e/src/leaflet-elevation.js#L131) function several times.

> The function that you indicated works great and that is why there is a flat line of heights from the end of one segment to the beginning of a new one. It is only necessary to paint over it :)) In general, it remains to deal with the paths in svg.

In your tests, you can **avoid editing library code** and access to the `<svg>` element as follows:

```js
var controlElevation = L.control.elevation( ... ); // leaflet.js chart object

// ...

var svg = d3.select(controlElevation.getContainer()).select("svg"); // d3.js object
var area = controlElevation._area;
var data = controlElevation._data;
```

> If I find a solution, I will definitely share it.

Ok thanks, if it can help try to check this pull too: https://github.com/MrMufflon/Leaflet.Elevation/pull/84

_Have a nice day,
Raruto_

**Comment by velocat** [comment]
> I can't see it in your code, but I think you solved this problem by calling the addData() function several times.


I did not write code so that there was no long text, but if interested, then here it is:

```
                         setElevationTrack = function(index) {
				var trace = traces[index+1]; // +1 i.e. 0=Main 
				controlElevation.clear();
				
				if(index == -1){				
					var Lines = trace.gpx.lines;
					Lines.forEach(function(line, i, arr){
						controlElevation.addData(line);
						controlElevation.hide(); //remove()
						line.setStyle({
							color: colors[countColor],
							weight: 4,
						});
						if(triggerDM == true){
							line.addDistanceMarkers();
						}else{
							line.removeDistanceMarkers();
						}
					});
					viewMain = true;
				}
				else{
					controlElevation.addData(trace.gpx.line);
					trace.gpx.setStyle({
						color: colors[countColor],
						weight: 4,
					}); 
					if(triggerDM == true){
						trace.gpx.line.addDistanceMarkers();
					}else{
						trace.gpx.line.removeDistanceMarkers();
					}
					viewMain = false;
				}	
				set_info(trace);
				mymap.fitBounds(trace.gpx.getBounds());	
			}

```

**Comment by Raruto** [comment]
Hi velocat,

I took a better look at the following [example](https://observablehq.com/@d3/area-with-missing-data):

![Screenshot_2020-09-30 Area with Missing Data(1)](https://user-images.githubusercontent.com/9614886/94667449-1e23fd80-030f-11eb-9fcf-da91c5a25acd.png)

Basically it works because where you have the `value: undefined`, the `date` value (eg. 2008-01-02) is still defined:

![Screenshot_2020-09-30 Area with Missing Data](https://user-images.githubusercontent.com/9614886/94667711-765aff80-030f-11eb-8fbc-957a9e26e658.png)

Unlike your dataset where you are also missing the `x` values:

```js
<gpx>
      <trk> .... </trk>
      <trk> .... </trk>
</gpx>
```

So, i think that to achieve a result like in the following [example](https://www.gpsvisualizer.com/tutorials/profiles_in_maps.html):

![Screenshot_2020-09-30 GPS Visualizer Dynamic elevation profiles in HTML maps](https://user-images.githubusercontent.com/9614886/94668271-203a8c00-0310-11eb-8416-d57fe1ab385f.png)

The [`data`](https://github.com/Raruto/leaflet-elevation/blob/2b35b0814cc778e9dec8f2d1f5a88bacdf5b0ae4/src/control.js#L116) property should be refactored into an "array of arrays" (or something like [that](https://github.com/GIScience/Leaflet.Heightgraph#supported-data))

As usual, pull requests are welcome...

_Have a nice day,
Raruto_

**Comment by carlos-mg89** [comment]
I've been doing some research, since multi segment tracks were working for me, and I've just realized that in version 1.5.6 (and below), multi segment tracks are working as expected. However, from version 1.6.0 and above, it doesn't work.

I guess that this change might have something to do with it: https://github.com/Raruto/leaflet-elevation/commit/0df53e92b6a8ba73bddd254bd59540d71ca1e47f#diff-3274f1a37032fb0ae4e2823def0007c634e869ae0dfc304ff6a12c36513c3a52L138-L172

I can see that the library stopped using the leaflet-gpx library. And that library has support for multi segment tracks.

Now they work like a charm. Do you think that it'll be possible to revert this change? I find it quite important, since there are a lot of GPX tracks with several segments in it. It's not about loading several tracks, it's just the same track, with different segments.

**Comment by velocat** [comment]
I can share my experience: 

I upload tracks via fileloader, even if they are on the server.  In it, as a handler, they use their own parser "importGeojson", in which I sort through the segments separately, create layers for them and add them to the array of segments.  

Next, I just switch between these layers, and each time I switch, I display the height profile for this segment, thereby always up-to-date information on the screen and there are no "gaps", as described above.  

If I want to show all the segment layers at the same time, I just disable the height profile.

![Segments](https://user-images.githubusercontent.com/534853/153562330-d063010a-4baa-4ffd-8cac-89bc5c9bcc38.gif)

**Comment by Raruto** [comment]
@velocat @carlos-mg89 if it interests you [here](https://github.com/Raruto/leaflet-elevation/tree/58c427ea87764ab62215f79750805f8eb71a0c3f/test) you can find what I have already tried to do in the past:

![multitrack](https://user-images.githubusercontent.com/9614886/153768130-401b66cd-9068-45c3-95b7-4731343b1c0e.png)

as you can see the geojson segments are already managed natively (the only problem is to find a way to split the chart profiles):

```js
var opts = {
	map: {
		center: [41.4583, 12.7059],
		zoom: 5,
		fullscreenControl: false,
		resizerControl: true,
		preferCanvas: true,
		rotate: true,
		bearing: 15,
		rotateControl: {
			closeOnZeroBearing: false
		},
	},
	elevationControl: {
		url: 'multi.gpx',
		options: {
			preferCanvas: false,
			theme: "lightblue-theme",
			collapsed: false,
			autohide: false,
			autofitBounds: true,
			position: "bottomleft",
			detached: true,
			summary: "inline",
			imperial: false,
			altitude: true, //"summary",
			distance: true, //"summary",
			slope: true, //"disabled",
			speed: "disabled",
			acceleration: "disabled",
			time: true, //"summary",
			timestamps: true,
			legend: true,
			followMarker: true,
			// zFollow: 13,
			ruler: true,
			// polyline: false,
			// marker: 'position-marker',
			// dragging: true,
			distanceMarkers: true,
			waypoints: true,
			wptIcons: true,
			wptLabels: true,
			almostOver: true,
		},
	},
	layersControl: {
		options: {
			collapsed: false,
		},
	},
};

var map              = new L.Map('map', opts.map);
var controlElevation = L.control.elevation(opts.elevationControl.options);
var controlLayer     = L.control.layers(null, null, opts.layersControl.options);

controlElevation.clear();
controlElevation.redraw();

map.on('eledata_loaded', function(e) {
	if (!controlLayer._map) controlLayer.addTo(map);
	controlElevation._gaps = controlElevation._data.reduce(function(array, item ) {
		if (item.trkEnd) array.push([]);
		if (array.length && (item.trkStart || item.trkEnd)) array[array.length-1].push(item);
		return array;
	}, []);
	e.layer.eachLayer((trkseg) => {
		if(trkseg.feature.geometry.type != "Point") {
			controlLayer.addOverlay(trkseg, trkseg.feature.properties.name);
		}
	});
	// console.log('eledata_loaded');
	// console.log(e);
});

controlElevation.on('eletrack_added', function(e) {
	this._data[0].trkStart = true;
	this._data[this._data.length-1].trkEnd = true;
	this.once('eledata_updated', function(e){
		this._data[this._data.length-1].trkStart = true;
	});
	// console.log('eletrack_added');
	// console.log(this._data[this._data.length-1], e);
});

controlElevation.load(opts.elevationControl.url);
controlElevation.addTo(map);
```
for those wishing to try, clone this repository into your localhost folder and inside move to the leaflet-elevation folder

- `git clone https://github.com/raruto/leaflet-elevation` (inside your localhost folder, eg. `/etc/www`)
- `cd leaflet-elevation`
- `npm i`
- `npm run dev`

After that you can start developing inside the `src` and `test` folders (open "http://localhost/leaflet-elevation/test" in your browser to see your changes).

_Have a nice day,
Raruto_

**Comment by Raruto** [comment]
Patch released in version **1.8.1**

![multi-track](https://user-images.githubusercontent.com/9614886/155755239-4d27aace-203c-40c9-91f7-d21039351b1b.png)

Below a simple example of how to use this in conjunction with L.Control.Layers:

```js
var map = new L.Map('map', opts.map);
var controlElevation = L.control.elevation(opts.elevationControl.options);
var controlLayer = L.control.layers(null, null, opts.layersControl.options);

// Optional
controlElevation.clear();
controlElevation.redraw();

// Add individual "multi-track" segments to layers control
map.on('eledata_loaded', function(e) {
  if (!controlLayer._map) controlLayer.addTo(map);
  e.layer.eachLayer((trkseg) => {
    if (trkseg.feature.geometry.type != "Point") {
      controlLayer.addOverlay(trkseg, trkseg.feature.properties.name);
    }
  });
});

controlElevation.load(opts.elevationControl.url);
controlElevation.addTo(map);
```

_Heartily,
Raruto_

**Comment by carlos-mg89** [comment]
@Raruto I tried with the 1.8.x versions but didn't make it work. There were some issues.

However, I've just tried it with the 1.9.6 release, and now it works like a charm.

Thanks a lot!!!
